lms/djangoapps/lti_provider/outcomes.py

```diff
@@ -0,0 +1,199 @@
+        if not result_service:
+            # TODO: There may be a way to recover from this error; if we know
+            # the LTI consumer that the request comes from then we may be able
+            # to figure out the result service URL. As it stands, though, this
+            # is a badly-formed LTI request
+            log.warn(
+                "Outcome Service: lis_outcome_service_url parameter missing "
+                "from scored assignment; we will be unable to return a score. "
+                "Request parameters: %s",
+                request_params
+            )
+            return
+
+        # We need to identify which tool consumer made the request. The best way
+        # to do this is to use the tool_consumer_instance_guid, which should be
+        # unique for each tool consumer. Unfortunately, that LTI parameter is
+        # optional, so we can't rely on it being available. In that case, we can
+        # fall back on the consumer key as a means of identifying the consumer.
+        # Note that this depends on every consumer being assigned a unique key
+        # (which is best practice anyway).
+        instance_guid = request_params.get('tool_consumer_instance_guid', None)
+        consumer_key = request_params.get('oauth_consumer_key', None)
+        assert consumer_key, "oauth_consumer_key is not an optional parameter."
+
+        # Both usage and course ID parameters are supplied in the LTI launch URL
+        usage_key = request_params.get('usage_key', None)
+        course_key = request_params.get('course_key', None)
+        assert None not in [usage_key, course_key], "usage_key ({}) and course_key ({}) should not be None".format(
+            usage_key, course_key
+        )
+
+        # Create a record of the outcome service if necessary
+        outcomes, __ = OutcomeService.objects.get_or_create(
+            lis_outcome_service_url=result_service,
+            consumer_key=consumer_key
+        )
+        # There could be a case where an earlier launch registered the outcome
+        # service but did not have the consumer instance guid available, in
+        # which case the outcome service would only have the consumer key
+        # available. If we now have better information, we can improve that
+        # record.
+        if instance_guid and not outcomes.instance_guid:
+            outcomes.instance_guid = instance_guid
+            outcomes.save()
+
+        # Create a record for this assignment. Note that there may already
+        # be a record, if this assignment has been launched more than once
+        # by the same user/tool consumer combination.
+        GradedAssignment.objects.get_or_create(
+            lis_result_sourcedid=result_id,
+            course_key=course_key,
+            usage_key=usage_key,
+            user=user,
+            outcome_service=outcomes
+        )
+
+
+# Pylint doesn't recognize members in the LXML module
+# pylint: disable=no-member
```

Comment:
Right I'm not saying do it on every line, just within the definition of the specific functions that need it. Eg instead of
```python
# pylint: disable=no-member
def f():
def f2():
def f3():
```

you do

```python
def f():
    # pylint: disable=no-member
def f2():
def f3():
    # pylint: disable=no-member
```

only disabling the warning for the functions that need it. That way if you add more functions to this file you're not erroneously ignoring a potentially real linting error.